packages/pigeon/lib/dart_generator.dart

```diff
@@ -263,17 +286,37 @@ void _writeFlutterApi(
   });
 }
 
-String _addGenericTypes(String dataType, String nullTag) {
-  switch (dataType) {
+/// Converts a [List] of [TypeDeclaration]s to a comma separated [String] to be
+/// used in Dart code.
+String _flattenTypeArguments(List<TypeDeclaration> args, String nullTag) {
+  return args
+      .map((TypeDeclaration arg) => arg.typeArguments == null
+          ? '${arg.dataType}$nullTag'
+          : '${arg.dataType}<${_flattenTypeArguments(arg.typeArguments!, nullTag)}>$nullTag')
+      .reduce((String value, String element) => '$value, $element');
+}
+
+/// Creates the type declaration for use in Dart code from a [NamedType] making sure
+/// that type arguments are used for primitive generic types.
+String _addGenericTypes(TypedEntity entity, String nullTag) {
+  switch (entity.dataType) {
     case 'List':
-      return 'List<Object$nullTag>$nullTag';
+      return (entity.typeArguments == null)
+          ? 'List<Object$nullTag>'
+          : 'List<${_flattenTypeArguments(entity.typeArguments!, nullTag)}>';
```

Comment:
> Here and below, consider a local for field.typeArguments so that you don't need the !. (You could even hoist it outside the switch and share the line; it's not always needed, but it's also just a trivial accessor so calling it when not needed isn't going to hurt anything.)

https://github.com/flutter/packages/pull/416#discussion_r681999800

Reply:
done